src/Workspaces/Core/Portable/Telemetry/TelemetryLogging.cs

```diff
@@ -3,26 +3,49 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using System.Threading;
+using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Telemetry;
 
 /// <summary>
 /// Provides access to posting telemetry events or adding information
-/// to aggregated telemetry events.
+/// to aggregated telemetry events. Posts pending telemetry at 30
+/// minute intervals.
 /// </summary>
 internal static class TelemetryLogging
 {
     private static ITelemetryLogProvider? s_logProvider;
+    private static AsyncBatchingWorkQueue? s_postTelemetryQueue;
 
     public const string KeyName = "Name";
     public const string KeyValue = "Value";
     public const string KeyLanguageName = "LanguageName";
     public const string KeyMetricName = "MetricName";
 
-    public static void SetLogProvider(ITelemetryLogProvider logProvider)
+    public static event EventHandler<EventArgs>? Flushed;
```

Comment:
what is the event for?

Reply:
RequestTelemetryLogger not only uses TelemetryLogging for aggregated telemetry, but also it's own accumulation of some counts. 

RequestTelemetryLogger uses Dispose to both notify this object to Flush and also handles firing the telemetry counts it is accumulating. However, Dispose isn't a reliable mechanism by which to fire telemetry, as the process might be terminated before we get the opportunity to act.

To handle that, the aggregating telemetry code previously fired off telemetry every 30 minutes, using an ABWQ. This PR moves that out to the TelemetryLogging level, but it's still not hooked into the telemetry that fires due to the accounting in RequestTelemetryLogger. 

This event allows RequestTelemetryLogger to hook into when that ABWQ derived telemetry firing is happening, and when it does RequestTelemetryLogger can fire off the telemetry that it is accounting for.